src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -337,47 +337,63 @@ void AnalyzeSignatureType(ImmutableArray<Location> locations, ITypeSymbol type)
+                        // 1. We're analyzing CoreLib and see the struct definition of a primitive type.
+                        // In all of these cases, the type does not have auto layout.
+                        // 2. We found a recursive type definition.
+                        // Recursive type definitions are invalid and Roslyn will emit another error diagnostic anyway,
+                        // so we don't care here.
+                        _isAutoLayoutOrContainsAutoLayoutCache.TryAdd(type, false);
+                        return false;
+                    }
 
-                            if (DiagnosticHelpers.TryConvertToUInt64(argument.Value, specialType, out ulong convertedLayoutKindValue) &&
-                                convertedLayoutKindValue == (ulong)LayoutKind.Auto)
+                    if (_structLayoutAttribute is not null)
+                    {
+                        foreach (var attr in type.GetAttributes(_structLayoutAttribute))
+                        {
+                            if (attr.ConstructorArguments.Length > 0
+                                && attr.ConstructorArguments[0] is TypedConstant argument
+                                && argument.Type is not null)
                             {
-                                _isAutoLayoutOrContainsAutoLayoutCache.TryAdd(type, true);
-                                return true;
+                                SpecialType specialType = argument.Type.TypeKind == TypeKind.Enum ?
+                                    ((INamedTypeSymbol)argument.Type).EnumUnderlyingType.SpecialType :
+                                    argument.Type.SpecialType;
+
+                                if (DiagnosticHelpers.TryConvertToUInt64(argument.Value, specialType, out ulong convertedLayoutKindValue) &&
+                                    convertedLayoutKindValue == (ulong)LayoutKind.Auto)
+                                {
+                                    _isAutoLayoutOrContainsAutoLayoutCache.TryAdd(type, true);
+                                    return true;
+                                }
                             }
                         }
                     }
-                }
 
-                foreach (var member in type.GetMembers())
-                {
-                    if (member is IFieldSymbol { IsStatic: false, Type.IsValueType: true } valueTypeField
-                        && TypeIsAutoLayoutOrContainsAutoLayout(valueTypeField.Type))
+                    foreach (var member in type.GetMembers())
                     {
-                        _isAutoLayoutOrContainsAutoLayoutCache.TryAdd(type, true);
-                        return true;
+                        if (member is IFieldSymbol { IsStatic: false, Type.IsValueType: true } valueTypeField
+                            && TypeIsAutoLayoutOrContainsAutoLayout(valueTypeField.Type, seenTypes.Add(type.OriginalDefinition)))
```

Comment:
This feels weird: we are repeatedly attempting to add same value to seenTypes in each iteration of the loop here.